### PR TITLE
metal_port datasource

### LINKED ISSUE #85: Create a metal_port datasource
Introducing `metal_port` as a datasource, ahead of any refactoring to offer a `metal_port` resource (https://github.com/equinix/terraform-provider-metal/issues/12#issuecomment-820444150) will give users the ability to work with ports that may be in use by appliances, connections, or devices.  The attributes of this datasource will give us a head start on a `metal_port` resource.

This data source could offer `device_id` search fields in addition to `id` (port_id).  Port name may also be a helpful search field. 

For users looking to fetch the `metal_device_network_type`  of a device, this would be possible by fetch the network type on each port and performing the same logic used by `metal_device_network_type` (this logic is not present in the EM API, this TF provider formulates it, like the EM Portal.  This single device "network_type" does not fit the broad reality of port configurations across all EM device models.

Users can get to the `device_id` through the existing `metal_vlan` data source.  Likewise, `ports` can be listed on the `metal_device` datasource.  However, there is no TF way (that I know of now) to get the ports related to a VLAN (or vice-versa) today.

This resource should follow `/ports/{id}`.  There may be other port attributes exposed in devices, connections, VLAN, or other resources, but my first inclination is to avoid including those in `metal_ports`. If we've omitted useful fields in those other resources, that are within a `ports` attribute, we should fill in the missing attributes in those resources.

### PR DESCRIPTION
**Pull request by t0mk**
fixes #85 

This PR adds a ds for metal_port, lookup of ports by ID or device ID and port name.

Signed-off-by: Tomas Karasek <tom.to.the.k@gmail.com>

### CODE REVIEW

File: docs/data-sources/port.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+## Argument Reference
+
+* `id` - (Required) ID of the port to read, conflicts with device_id.
+* `device_id` - (Required) 4 or 6, depending on which block you are looking for.
+* `port_name` - (Required) Whether to look for public or private block.
+
+## Attributes Reference
+
+* `bond_port` - Whether port is a bond port or physical port
```
Comment (displague) [inline docs/data-sources/port.md:44]: `bond_port` here is a TF boolean based on the `type` field matching `NeworkBondPort`.  I think we should just expose `type` and let the user apply this logic if they need it.  The `type` field could present new options in the future.

There are a few more attributes that we can expose here: https://github.com/packethost/packngo/blob/master/ports.go#L45-L75
* `type`
* `data` { mac, bonded }
* `disbond_operation_supported`
* `native_virtual_network`
* `virtual_networks`
* `bond` { id, name }

File: docs/data-sources/port.md [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    device_id        = metal_device.test.id
+    port_name        = "eth0"
+}
+```
+
+## Argument Reference
+
+* `id` - (Required) ID of the port to read, conflicts with device_id.
+* `device_id` - (Required) 4 or 6, depending on which block you are looking for.
+* `port_name` - (Required) Whether to look for public or private block.
```
Comment (displague) [inline docs/data-sources/port.md:40]: Perhaps we should use the name `name`, to match the API.

File: docs/data-sources/port.md [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+data "metal_port" "test" {
+    device_id        = metal_device.test.id
+    port_name        = "eth0"
+}
+```
+
+## Argument Reference
+
+* `id` - (Required) ID of the port to read, conflicts with device_id.
+* `device_id` - (Required) 4 or 6, depending on which block you are looking for.
```
Comment (displague) [inline docs/data-sources/port.md:39]: I considered suggesting that we call this `hardware` to match the API, but I'm not sure how we would be able to determine if the user means "hardware is a reservation" or "hardware is a device instance" or "hardware is something else".

It makes sense to use "device_id" + "port.name" for lookups, and if we need more lookups in the future we'll have to add more lookup fields.

Should we set `device_id` based on the `hardware.href` returned (assuming there href basename is the id)?  Perhaps we need to query `/ports/{id}?include=hardware` to gather the needed information.

File: docs/data-sources/port.md (hunk 1 above)
Comment (displague) [reply docs/data-sources/port.md:44]: Other concerns aside, I'd be glad to merge this sooner and add those fields in a follow-up.

File: metal/datasource_metal_port_test.go [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+	return fmt.Sprintf(`
+
+resource "metal_project" "test" {
+    name = "tfacc-port-%s"
+}
+
+resource "metal_device" "test" {
+  hostname         = "tfacc-test-device-port"
+  plan             = "c3.medium.x86"
+  facilities       = ["sv15"]
```
Comment (displague) [inline metal/datasource_metal_port_test.go:40]: Unless we are testing `facility` workflows, we should test with metros for a better success rate in tests (more availability).

File: docs/data-sources/port.md (hunk 3 above)
Comment (t0mk) [reply docs/data-sources/port.md:39]: sorry about the doc string for the `device_id`.

I'm a bit confused here. The hardware ID is different than the device ID, and the href link in hardware is not GETtable. I get 404 getting e..g /hardware/8eec90d7-62c4-47c2-b541-b0a0a3f1c428.

I can see the hardware resource when I GET "/ports/9ea27e03-8c5f-4a2a-ac6e-466cb1076c36?include=hardware". It has the device in `instance` attribute. However, we don't have a struct for the "hardware" resource here or in packngo.

File: docs/data-sources/port.md (hunk 2 above)
Comment (t0mk) [reply docs/data-sources/port.md:40]: ok

File: metal/datasource_metal_port_test.go (hunk 4 above)
Comment (t0mk) [reply metal/datasource_metal_port_test.go:40]: :+1:

**Pull request merged**